File: Submissions/213_House_Robber_II.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if(n<=2) return rob(nums, 0, n-1);
        return max(rob(nums, 0, n-2), rob(nums, 1, n-1));
    }
    
    int rob(vector<int>& nums, int begin, int end) {
        int prev1 = 0, prev2 = 0, curr = 0;
        for(int i = begin; i <= end; ++i){
            curr = max(prev1 + nums[i], prev2); 
            prev1 = prev2;
            prev2 = curr;
        }
        
        return curr;
    }
};

File: Kotlin/213_House_Robber_II.kt
class Solution {
    fun rob(nums: IntArray): Int {
        val n = nums.size
        if(n <= 2) return recursive(nums, 0, n-1)
        return maxOf(recursive(nums, 0, n-2), recursive(nums, 1, n-1))
    }
    
    fun recursive(nums: IntArray, begin: Int, end: Int): Int{
        var prev1 = 0
        var prev2 = 0
        var curr = 0
        
        for(i in begin..end){
            curr = maxOf(prev1 + nums[i], prev2)
            prev1 = prev2
            prev2 = curr
        }
        
        return curr
    }
}